File: blueprints/workspace/app/web/handler/api/comment_test.go
package api_test

import (
	"net/http"
	"testing"

	"github.com/go-mizu/blueprints/workspace/feature/comments"
)

func TestCommentCreate(t *testing.T) {
	ts := NewTestServer(t)
	defer ts.Close()

	// Setup
	_, cookie := ts.Register("commentcreate@example.com", "Comment Create", "password123")
	ws := createTestWorkspace(ts, cookie, "Comment Workspace", "comment-ws")
	page := createTestPage(ts, cookie, ws.ID, "Comment Test Page")

	tests := []struct {
		name       string
		body       map[string]interface{}
		wantStatus int
	}{
		{
			name: "page comment",
			body: map[string]interface{}{
				"workspace_id": ws.ID,
				"target_type":  "page",
				"target_id":    page.ID,
				"content": []map[string]interface{}{
					{"type": "text", "text": "This is a comment"},
				},
			},
			wantStatus: http.StatusCreated,
		},
		{
			name: "comment with formatting",
			body: map[string]interface{}{
				"workspace_id": ws.ID,
				"target_type":  "page",
				"target_id":    page.ID,
				"content": []map[string]interface{}{
					{
						"type": "text",
						"text": "Bold comment",
						"annotations": map[string]interface{}{
							"bold": true,
						},
					},
				},
			},
			wantStatus: http.StatusCreated,
		},
		{
			name: "missing target",
			body: map[string]interface{}{
				"content": []map[string]interface{}{
					{"type": "text", "text": "No target"},
				},
			},
			wantStatus: http.StatusCreated, // App allows comments without target
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			resp := ts.Request("POST", "/api/v1/comments", tt.body, cookie)
			ts.ExpectStatus(resp, tt.wantStatus)

			if tt.wantStatus == http.StatusCreated {
				var comment comments.Comment
				ts.ParseJSON(resp, &comment)

				if comment.ID == "" {
					t.Error("comment ID should not be empty")
				}
				// Only check target_id if target_id was provided in request
				if tt.body["target_id"] != nil && comment.TargetID != page.ID {
					t.Errorf("target_id = %q, want %q", comment.TargetID, page.ID)
				}
			}
			resp.Body.Close()
		})
	}
}

func TestCommentCreateReply(t *testing.T) {
	ts := NewTestServer(t)
	defer ts.Close()

	// Setup
	_, cookie := ts.Register("commentreply@example.com", "Comment Reply", "password123")
	ws := createTestWorkspace(ts, cookie, "Reply Workspace", "reply-ws")
	page := createTestPage(ts, cookie, ws.ID, "Reply Test Page")

	// Create parent comment
	resp := ts.Request("POST", "/api/v1/comments", map[string]interface{}{
		"workspace_id": ws.ID,
		"target_type":  "page",
		"target_id":    page.ID,
		"content": []map[string]interface{}{
			{"type": "text", "text": "Parent comment"},
		},
	}, cookie)
	ts.ExpectStatus(resp, http.StatusCreated)

	var parent comments.Comment
	ts.ParseJSON(resp, &parent)

	// Create reply
	resp = ts.Request("POST", "/api/v1/comments", map[string]interface{}{
		"workspace_id": ws.ID,
		"target_type":  "page",
		"target_id":    page.ID,
		"parent_id":    parent.ID,
		"content": []map[string]interface{}{
			{"type": "text", "text": "Reply comment"},
		},
	}, cookie)
	ts.ExpectStatus(resp, http.StatusCreated)

	var reply comments.Comment
	ts.ParseJSON(resp, &reply)

	if reply.ParentID != parent.ID {
		t.Errorf("parent_id = %q, want %q", reply.ParentID, parent.ID)
	}
}

func TestCommentUpdate(t *testing.T) {
	ts := NewTestServer(t)
	defer ts.Close()

	// Setup
	_, cookie := ts.Register("commentupdate@example.com", "Comment Update", "password123")
	ws := createTestWorkspace(ts, cookie, "Update Workspace", "update-comment-ws")
	page := createTestPage(ts, cookie, ws.ID, "Update Test Page")

	// Create comment
	resp := ts.Request("POST", "/api/v1/comments", map[string]interface{}{
		"workspace_id": ws.ID,
		"target_type":  "page",
		"target_id":    page.ID,
		"content": []map[string]interface{}{
			{"type": "text", "text": "Original comment"},
		},
	}, cookie)
	ts.ExpectStatus(resp, http.StatusCreated)

	var created comments.Comment
	ts.ParseJSON(resp, &created)

	t.Run("update content", func(t *testing.T) {
		resp := ts.Request("PATCH", "/api/v1/comments/"+created.ID, map[string]interface{}{
			"content": []map[string]interface{}{
				{"type": "text", "text": "Updated comment"},
			},
		}, cookie)
		ts.ExpectStatus(resp, http.StatusOK)

		var updated comments.Comment
		ts.ParseJSON(resp, &updated)

		if len(updated.Content) == 0 || updated.Content[0].Text != "Updated comment" {
			t.Error("content should be updated")
		}
	})

	t.Run("non-existent comment", func(t *testing.T) {
		resp := ts.Request("PATCH", "/api/v1/comments/non-existent-id", map[string]interface{}{
			"content": []map[string]interface{}{
				{"type": "text", "text": "Update"},
			},
		}, cookie)
		ts.ExpectStatus(resp, http.StatusInternalServerError)
		resp.Body.Close()
	})
}

func TestCommentDelete(t *testing.T) {
	ts := NewTestServer(t)
	defer ts.Close()

	// Setup
	_, cookie := ts.Register("commentdelete@example.com", "Comment Delete", "password123")
	ws := createTestWorkspace(ts, cookie, "Delete Workspace", "delete-comment-ws")
	page := createTestPage(ts, cookie, ws.ID, "Delete Test Page")

	// Create comment
	resp := ts.Request("POST", "/api/v1/comments", map[string]interface{}{
		"workspace_id": ws.ID,
		"target_type":  "page",
		"target_id":    page.ID,
		"content": []map[string]interface{}{
			{"type": "text", "text": "To delete"},
		},
	}, cookie)
	ts.ExpectStatus(resp, http.StatusCreated)

	var created comments.Comment
	ts.ParseJSON(resp, &created)

	t.Run("delete comment", func(t *testing.T) {
		resp := ts.Request("DELETE", "/api/v1/comments/"+created.ID, nil, cookie)
		ts.ExpectStatus(resp, http.StatusOK)
		resp.Body.Close()
	})
}

func TestCommentList(t *testing.T) {
	ts := NewTestServer(t)
	defer ts.Close()

	// Setup
	_, cookie := ts.Register("commentlist@example.com", "Comment List", "password123")
	ws := createTestWorkspace(ts, cookie, "List Workspace", "list-comment-ws")
	page := createTestPage(ts, cookie, ws.ID, "List Test Page")

	// Create comments
	for i := 0; i < 3; i++ {
		resp := ts.Request("POST", "/api/v1/comments", map[string]interface{}{
			"workspace_id": ws.ID,
			"target_type":  "page",
			"target_id":    page.ID,
			"content": []map[string]interface{}{
				{"type": "text", "text": "Comment " + string(rune('A'+i))},
			},
		}, cookie)
		ts.ExpectStatus(resp, http.StatusCreated)
		resp.Body.Close()
	}

	t.Run("list page comments", func(t *testing.T) {
		resp := ts.Request("GET", "/api/v1/pages/"+page.ID+"/comments", nil, cookie)
		ts.ExpectStatus(resp, http.StatusOK)

		var commentList []*comments.Comment
		ts.ParseJSON(resp, &commentList)

		if len(commentList) < 3 {
			t.Errorf("expected at least 3 comments, got %d", len(commentList))
		}
	})
}

func TestCommentResolve(t *testing.T) {
	ts := NewTestServer(t)
	defer ts.Close()

	// Setup
	_, cookie := ts.Register("commentresolve@example.com", "Comment Resolve", "password123")
	ws := createTestWorkspace(ts, cookie, "Resolve Workspace", "resolve-ws")
	page := createTestPage(ts, cookie, ws.ID, "Resolve Test Page")

	// Create comment
	resp := ts.Request("POST", "/api/v1/comments", map[string]interface{}{
		"workspace_id": ws.ID,
		"target_type":  "page",
		"target_id":    page.ID,
		"content": []map[string]interface{}{
			{"type": "text", "text": "To resolve"},
		},
	}, cookie)
	ts.ExpectStatus(resp, http.StatusCreated)

	var created comments.Comment
	ts.ParseJSON(resp, &created)

	if created.IsResolved {
		t.Error("new comment should not be resolved")
	}

	t.Run("resolve comment", func(t *testing.T) {
		resp := ts.Request("POST", "/api/v1/comments/"+created.ID+"/resolve", nil, cookie)
		ts.ExpectStatus(resp, http.StatusOK)
		resp.Body.Close()
	})
}

func TestCommentThreaded(t *testing.T) {
	ts := NewTestServer(t)
	defer ts.Close()

	// Setup
	_, cookie := ts.Register("commentthread@example.com", "Comment Thread", "password123")
	ws := createTestWorkspace(ts, cookie, "Thread Workspace", "thread-ws")
	page := createTestPage(ts, cookie, ws.ID, "Thread Test Page")

	// Create parent comment
	resp := ts.Request("POST", "/api/v1/comments", map[string]interface{}{
		"workspace_id": ws.ID,
		"target_type":  "page",
		"target_id":    page.ID,
		"content": []map[string]interface{}{
			{"type": "text", "text": "Thread starter"},
		},
	}, cookie)
	ts.ExpectStatus(resp, http.StatusCreated)

	var parent comments.Comment
	ts.ParseJSON(resp, &parent)

	// Create multiple replies
	for i := 0; i < 3; i++ {
		resp := ts.Request("POST", "/api/v1/comments", map[string]interface{}{
			"workspace_id": ws.ID,
			"target_type":  "page",
			"target_id":    page.ID,
			"parent_id":    parent.ID,
			"content": []map[string]interface{}{
				{"type": "text", "text": "Reply " + string(rune('1'+i))},
			},
		}, cookie)
		ts.ExpectStatus(resp, http.StatusCreated)
		resp.Body.Close()
	}

	// List all comments - replies are nested in parent's Replies field
	resp = ts.Request("GET", "/api/v1/pages/"+page.ID+"/comments", nil, cookie)
	ts.ExpectStatus(resp, http.StatusOK)

	var commentList []*comments.Comment
	ts.ParseJSON(resp, &commentList)

	// Should have 1 root comment with 3 replies nested inside
	if len(commentList) != 1 {
		t.Errorf("expected 1 root comment, got %d", len(commentList))
	}
	if len(commentList) > 0 && len(commentList[0].Replies) != 3 {
		t.Errorf("expected 3 replies, got %d", len(commentList[0].Replies))
	}
}

func TestCommentWithMention(t *testing.T) {
	ts := NewTestServer(t)
	defer ts.Close()

	// Setup - create two users
	user1, cookie1 := ts.Register("mentioner@example.com", "Mentioner", "password123")
	user2, _ := ts.Register("mentioned@example.com", "Mentioned", "password123")

	ws := createTestWorkspace(ts, cookie1, "Mention Workspace", "mention-ws")
	page := createTestPage(ts, cookie1, ws.ID, "Mention Test Page")

	// Create comment with mention
	resp := ts.Request("POST", "/api/v1/comments", map[string]interface{}{
		"workspace_id": ws.ID,
		"target_type":  "page",
		"target_id":    page.ID,
		"content": []map[string]interface{}{
			{"type": "text", "text": "Hey "},
			{
				"type": "text",
				"text": "@Mentioned",
				"mention": map[string]interface{}{
					"type":    "user",
					"user_id": user2.ID,
				},
			},
			{"type": "text", "text": " check this out!"},
		},
	}, cookie1)
	ts.ExpectStatus(resp, http.StatusCreated)

	var comment comments.Comment
	ts.ParseJSON(resp, &comment)

	if comment.AuthorID != user1.ID {
		t.Errorf("author_id = %q, want %q", comment.AuthorID, user1.ID)
	}
}

func TestCommentUnauthenticated(t *testing.T) {
	ts := NewTestServer(t)
	defer ts.Close()

	// Setup
	_, cookie := ts.Register("commentauth@example.com", "Comment Auth", "password123")
	ws := createTestWorkspace(ts, cookie, "Auth Workspace", "auth-comment-ws")
	page := createTestPage(ts, cookie, ws.ID, "Auth Test Page")

	// Create comment
	resp := ts.Request("POST", "/api/v1/comments", map[string]interface{}{
		"workspace_id": ws.ID,
		"target_type":  "page",
		"target_id":    page.ID,
		"content": []map[string]interface{}{
			{"type": "text", "text": "Auth test"},
		},
	}, cookie)
	ts.ExpectStatus(resp, http.StatusCreated)

	var comment comments.Comment
	ts.ParseJSON(resp, &comment)

	tests := []struct {
		name   string
		method string
		path   string
	}{
		{"create comment", "POST", "/api/v1/comments"},
		{"update comment", "PATCH", "/api/v1/comments/" + comment.ID},
		{"delete comment", "DELETE", "/api/v1/comments/" + comment.ID},
		{"list comments", "GET", "/api/v1/pages/" + page.ID + "/comments"},
		{"resolve comment", "POST", "/api/v1/comments/" + comment.ID + "/resolve"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			resp := ts.Request(tt.method, tt.path, nil) // No cookie
			ts.ExpectStatus(resp, http.StatusUnauthorized)
			resp.Body.Close()
		})
	}
}

File: blueprints/workspace/app/backend/hono/src/tests/e2e/comment.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { createTestApp, withSession, setupTestContext, createPage, createBlock, registerUser } from './setup';
import type { TestApp } from './setup';

describe('Comment E2E Tests', () => {
  let app: TestApp;

  beforeEach(() => {
    app = createTestApp();
  });

  describe('POST /api/v1/comments', () => {
    it('should create a page comment', async () => {
      const { sessionId, workspace, user } = await setupTestContext(app);
      const page = await createPage(app, sessionId, workspace.id);

      const res = await app.request('/api/v1/comments', {
        method: 'POST',
        headers: withSession({ 'Content-Type': 'application/json' }, sessionId),
        body: JSON.stringify({
          workspaceId: workspace.id,
          targetType: 'page',
          targetId: page.id,
          content: [{ type: 'text', text: { content: 'Great page!' } }],
        }),
      });

      expect(res.status).toBe(201);
      const data = await res.json() as { comment: { targetType: string; targetId: string; authorId: string } };
      expect(data.comment.targetType).toBe('page');
      expect(data.comment.targetId).toBe(page.id);
      expect(data.comment.authorId).toBe(user.id);
    });

    it('should create a block comment', async () => {
      const { sessionId, workspace } = await setupTestContext(app);
      const page = await createPage(app, sessionId, workspace.id);
      const block = await createBlock(app, sessionId, page.id);

      const res = await app.request('/api/v1/comments', {
        method: 'POST',
        headers: withSession({ 'Content-Type': 'application/json' }, sessionId),
        body: JSON.stringify({
          workspaceId: workspace.id,
          targetType: 'block',
          targetId: block.id,
          content: [{ type: 'text', text: { content: 'Comment on block' } }],
        }),
      });

      expect(res.status).toBe(201);
      const data = await res.json() as { comment: { targetType: string; targetId: string } };
      expect(data.comment.targetType).toBe('block');
      expect(data.comment.targetId).toBe(block.id);
    });

    it('should create a reply to comment', async () => {
      const { sessionId, workspace } = await setupTestContext(app);
      const page = await createPage(app, sessionId, workspace.id);

      // Create parent comment
      const parentRes = await app.request('/api/v1/comments', {
        method: 'POST',
        headers: withSession({ 'Content-Type': 'application/json' }, sessionId),
        body: JSON.stringify({
          workspaceId: workspace.id,
          targetType: 'page',
          targetId: page.id,
          content: [{ type: 'text', text: { content: 'Parent comment' } }],
        }),
      });
      const parentComment = (await parentRes.json() as { comment: { id: string } }).comment;

      // Create reply
      const res = await app.request('/api/v1/comments', {
        method: 'POST',
        headers: withSession({ 'Content-Type': 'application/json' }, sessionId),
        body: JSON.stringify({
          workspaceId: workspace.id,
          targetType: 'page',
          targetId: page.id,
          parentId: parentComment.id,
          content: [{ type: 'text', text: { content: 'Reply comment' } }],
        }),
      });

      expect(res.status).toBe(201);
      const data = await res.json() as { comment: { parentId: string } };
      expect(data.comment.parentId).toBe(parentComment.id);
    });
  });

  describe('GET /api/v1/pages/:id/comments', () => {
    it('should list page comments', async () => {
      const { sessionId, workspace } = await setupTestContext(app);
      const page = await createPage(app, sessionId, workspace.id);

      // Create comments
      await app.request('/api/v1/comments', {
        method: 'POST',
        headers: withSession({ 'Content-Type': 'application/json' }, sessionId),
        body: JSON.stringify({
          workspaceId: workspace.id,
          targetType: 'page',
          targetId: page.id,
          content: [{ type: 'text', text: { content: 'Comment 1' } }],
        }),
      });
      await app.request('/api/v1/comments', {
        method: 'POST',
        headers: withSession({ 'Content-Type': 'application/json' }, sessionId),
        body: JSON.stringify({
          workspaceId: workspace.id,
          targetType: 'page',
          targetId: page.id,
          content: [{ type: 'text', text: { content: 'Comment 2' } }],
        }),
      });

      const res = await app.request(`/api/v1/pages/${page.id}/comments`, {
        headers: withSession({}, sessionId),
      });

      expect(res.status).toBe(200);
      const data = await res.json() as { comments: any[] };
      expect(data.comments.length).toBe(2);
    });

    it('should return comments with author info', async () => {
      const { sessionId, workspace, user } = await setupTestContext(app);
      const page = await createPage(app, sessionId, workspace.id);

      await app.request('/api/v1/comments', {
        method: 'POST',
        headers: withSession({ 'Content-Type': 'application/json' }, sessionId),
        body: JSON.stringify({
          workspaceId: workspace.id,
          targetType: 'page',
          targetId: page.id,
          content: [{ type: 'text', text: { content: 'My comment' } }],
        }),
      });

      const res = await app.request(`/api/v1/pages/${page.id}/comments`, {
        headers: withSession({}, sessionId),
      });

      expect(res.status).toBe(200);
      const data = await res.json() as { comments: { authorId: string }[] };
      expect(data.comments[0].authorId).toBe(user.id);
    });
  });

  describe('PATCH /api/v1/comments/:id', () => {
    it('should update comment content', async () => {
      const { sessionId, workspace } = await setupTestContext(app);
      const page = await createPage(app, sessionId, workspace.id);

      const commentRes = await app.request('/api/v1/comments', {
        method: 'POST',
        headers: withSession({ 'Content-Type': 'application/json' }, sessionId),
        body: JSON.stringify({
          workspaceId: workspace.id,
          targetType: 'page',
          targetId: page.id,
          content: [{ type: 'text', text: { content: 'Original' } }],
        }),
      });
      const comment = (await commentRes.json() as { comment: { id: string } }).comment;

      const res = await app.request(`/api/v1/comments/${comment.id}`, {
        method: 'PATCH',
        headers: withSession({ 'Content-Type': 'application/json' }, sessionId),
        body: JSON.stringify({
          content: [{ type: 'text', text: { content: 'Updated' } }],
        }),
      });

      expect(res.status).toBe(200);
      const data = await res.json() as { comment: { content: { text: string }[] } };
      expect(data.comment.content[0].text).toBe('Updated');
    });

    it('should not allow non-author to update', async () => {
      const { sessionId: authorSession, workspace } = await setupTestContext(app);
      const page = await createPage(app, authorSession, workspace.id);

      const commentRes = await app.request('/api/v1/comments', {
        method: 'POST',
        headers: withSession({ 'Content-Type': 'application/json' }, authorSession),
        body: JSON.stringify({
          workspaceId: workspace.id,
          targetType: 'page',
          targetId: page.id,
          content: [{ type: 'text', text: { content: 'Original' } }],
        }),
      });
      const comment = (await commentRes.json() as { comment: { id: string } }).comment;

      // Add another user to workspace and try to update
      const { user: otherUser, sessionId: otherSession } = await registerUser(app);
      await app.request(`/api/v1/workspaces/${workspace.id}/members`, {
        method: 'POST',
        headers: withSession({ 'Content-Type': 'application/json' }, authorSession),
        body: JSON.stringify({ userId: otherUser.id, role: 'member' }),
      });

      const res = await app.request(`/api/v1/comments/${comment.id}`, {
        method: 'PATCH',
        headers: withSession({ 'Content-Type': 'application/json' }, otherSession),
        body: JSON.stringify({
          content: [{ type: 'text', text: { content: 'Hacked' } }],
        }),
      });

      expect(res.status).toBe(403);
    });
  });

  describe('DELETE /api/v1/comments/:id', () => {
    it('should delete comment by author', async () => {
      const { sessionId, workspace } = await setupTestContext(app);
      const page = await createPage(app, sessionId, workspace.id);

      const commentRes = await app.request('/api/v1/comments', {
        method: 'POST',
        headers: withSession({ 'Content-Type': 'application/json' }, sessionId),
        body: JSON.stringify({
          workspaceId: workspace.id,
          targetType: 'page',
          targetId: page.id,
          content: [{ type: 'text', text: { content: 'To delete' } }],
        }),
      });
      const comment = (await commentRes.json() as { comment: { id: string } }).comment;

      const res = await app.request(`/api/v1/comments/${comment.id}`, {
        method: 'DELETE',
        headers: withSession({}, sessionId),
      });

      expect(res.status).toBe(200);
    });

    it('should not allow non-author to delete', async () => {
      const { sessionId: authorSession, workspace } = await setupTestContext(app);
      const page = await createPage(app, authorSession, workspace.id);

      const commentRes = await app.request('/api/v1/comments', {
        method: 'POST',
        headers: withSession({ 'Content-Type': 'application/json' }, authorSession),
        body: JSON.stringify({
          workspaceId: workspace.id,
          targetType: 'page',
          targetId: page.id,
          content: [{ type: 'text', text: { content: 'Protected' } }],
        }),
      });
      const comment = (await commentRes.json() as { comment: { id: string } }).comment;

      const { user: otherUser, sessionId: otherSession } = await registerUser(app);
      await app.request(`/api/v1/workspaces/${workspace.id}/members`, {
        method: 'POST',
        headers: withSession({ 'Content-Type': 'application/json' }, authorSession),
        body: JSON.stringify({ userId: otherUser.id, role: 'member' }),
      });

      const res = await app.request(`/api/v1/comments/${comment.id}`, {
        method: 'DELETE',
        headers: withSession({}, otherSession),
      });

      expect(res.status).toBe(403);
    });
  });

  describe('POST /api/v1/comments/:id/resolve', () => {
    it('should resolve comment thread', async () => {
      const { sessionId, workspace } = await setupTestContext(app);
      const page = await createPage(app, sessionId, workspace.id);

      const commentRes = await app.request('/api/v1/comments', {
        method: 'POST',
        headers: withSession({ 'Content-Type': 'application/json' }, sessionId),
        body: JSON.stringify({
          workspaceId: workspace.id,
          targetType: 'page',
          targetId: page.id,
          content: [{ type: 'text', text: { content: 'Issue found' } }],
        }),
      });
      const comment = (await commentRes.json() as { comment: { id: string } }).comment;

      const res = await app.request(`/api/v1/comments/${comment.id}/resolve`, {
        method: 'POST',
        headers: withSession({}, sessionId),
      });

      expect(res.status).toBe(200);
      const data = await res.json() as { comment: { isResolved: boolean } };
      expect(data.comment.isResolved).toBe(true);
    });
  });
});
